docs/administration/grid-data-sources.md

```diff
@@ -40,14 +42,29 @@ $queryBuilder = $transportRepository->getQueryBuilderForAll()
 $dataSource = new QueryBuilderWithRowManipulatorDataSource(
     $queryBuilder,
     't.id',
-    function ($row) {
-        $transport = $transportRepository->findById($row['t']['id']);
+    function ($row, $results) {
+        if($this->transportsIndexedByLocalCache === null) {
+            $this->transportsIndexedByIdLocalCache = $transportRepository->findAllByIdsIndexedById(array_map(fn ($result) => $result['t']['id'], $results));
+        }
+
+        $transport = $this->transportsIndexedByIdLocalCache[$row['t']['id']];
         $row['displayPrice'] = getDisplayPrice($transport);
         return $row;
-    }
+    },
+    SortableNullsWalker::class
 );
 ```
 
+As in the case of QueryBuilderDataSource, it is possible to change the setting of the $hint parameter, see QueryBuilderDataSource.
+
+### Tips for optimal usage QueryBuilderDataSource or QueryBuilderWithRowManipulatorDataSource
+
+- The query should return an array of scalar values instead of an entity if the table contains large collections of records.
+- Grid with callback calls the callback for each item separately. To avoid calling SQL queries for each item, use the second parameter of the callback function in which the entire result for the bulk query is stored in the local cache. See. OrderController.
```

Comment:
```suggestion
- Grid with callback calls the callback for each item separately. To avoid calling SQL queries for each item, use the second parameter of the callback function in which the entire result for the bulk query is stored in the local cache. See e.g. `Shopsys\FrameworkBundle\Controller\Admin\OrderController::getOrdersGrid()`.
```